cmd/vault.go

```diff
@@ -397,3 +402,326 @@ func readJSON(name string) (secret interface{}, err error) {
+}
+
+//
+
+// create user secret
+func userSecretCreateCmd() *cobra.Command {
+	options := vaultOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "create",
+		Short: "Create new user secret",
+		Long:  `Create new user secret`,
+		Example: `
+	privx-cli user-secrets create [access flags] --name <SECRET-NAME>
+		--allow-read-to <ROLE-ID>
+		--allow-write-to <ROLE-ID>
+		...
+		JSON-FILE
+		`,
+		Args:         cobra.ExactArgs(1),
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return userSecretCreate(args, options)
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.secretName, "name", "", "secret name")
+	flags.StringArrayVar(&options.vaultReadTo, "allow-read-to", []string{}, "read by role ID")
+	flags.StringArrayVar(&options.vaultWriteTo, "allow-write-to", []string{}, "write by role ID")
+	flags.StringVar(&options.ownerID, "owner-id", "", "secret's owner ID")
+	cmd.MarkFlagRequired("owner-id")
+	cmd.MarkFlagRequired("name")
+	cmd.MarkFlagRequired("read-by")
+	cmd.MarkFlagRequired("write-by")
+
+	return cmd
+}
+
+func userSecretCreate(args []string, options vaultOptions) error {
+	secret, err := readJSON(args[0])
+	if err != nil {
+		return err
+	}
+
+	secretID := &vault.SecretID{
+		OwnerID: options.ownerID,
+		Name:    options.secretName,
+	}
+	api := vault.New(curl())
+	if err := api.CreateUserSecret(*secretID, options.vaultReadTo,
+		options.vaultWriteTo, secret); err != nil {
+		return err
+	}
+
+	return stdout(secret)
+}
+
+//
+
+// update user secret
+func userSecretUpdateCmd() *cobra.Command {
+	options := vaultOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "update",
+		Short: "Update user secret",
+		Long:  `Update existing user secret`,
+		Example: `
+	privx-cli user-secrets update [access flags] --owner-id <SECRET-NAME> --name <SECRET-NAME> JSON-FILE
+	privx-cli user-secrets update [access flags] --owner-id <SECRET-NAME> --name <SECRET-NAME>
+		--allow-read-to <ROLE-ID>
+		--allow-write-to <ROLE-ID>
+		...
+		JSON-FILE
+		`,
+		Args:         cobra.ExactArgs(1),
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return userSecretUpdate(options, args)
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.secretName, "name", "", "secret name")
+	flags.StringArrayVar(&options.vaultReadTo, "allow-read-to", []string{}, "read by role ID")
+	flags.StringArrayVar(&options.vaultWriteTo, "allow-write-to", []string{}, "write by role ID")
+	flags.StringVar(&options.ownerID, "owner-id", "", "secret's owner ID")
+	cmd.MarkFlagRequired("owner-id")
+	cmd.MarkFlagRequired("name")
+
+	return cmd
+}
+
+func userSecretUpdate(options vaultOptions, args []string) error {
+	secret, err := readJSON(args[0])
+	if err != nil {
+		return err
+	}
+
+	secretID := &vault.SecretID{
+		OwnerID: options.ownerID,
+		Name:    options.secretName,
+	}
+	api := vault.New(curl())
+	bag, err := api.UserSecret(*secretID)
+	if err != nil {
+		return err
+	}
+
+	if len(options.vaultReadTo) == 0 {
```

Comment:
Here there is an issue: does the empty option string mean "clear all current roles" or "do not change what there is already". Now, regardless of the answer to the question the next one would be: "if empty array is used for \<insert your preferred interpretation\> then how the other case can be indicated?"

Obviously empty array cannot indicate both "no change please" and "clear everything please". With the current implementation, how would we clear read/write roles? there seems to be no way of doing it. 

Thus, there is a need to maintain 3 different invocations:
```
// Invocation with value
// Role: ["a", "b", "c"]
cmd --role=a,b,c

// Invocation without value (i.e use NoOptDefVal)
// Role: [], This will result in all roles being cleared
cmd --role

// Invocation without flag (i.e use default value)
// Role: nil slice -- Do not change the roles
cmd
```

An alternative solution is to assume that the roles passed to the function are in the final desired state. That is we pass the roles directly to the SDK, and thus the role flags are required. The function itself should not fetch existing metadata. This agrees with the current implementation of the SDK and trivial to do albeit it is very inconvenient to use.

To keep the PR from exploding the Update command can be left out from this PR. While implement this function It is best to also update the SDK such that mkVaultReq() also check if the slice passed to it is nil or not and act accordingly. Currently the SDK maintains that the slice cannot be nil. So go-client needs to fetch secret metadata even though no change to the metadata is designed.